libcxx/test/libcxx/containers/strings/basic.string/asan_turning_off.pass.cpp

```diff
@@ -0,0 +1,78 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// REQUIRES: asan
+// UNSUPPORTED: c++03
+
+// <string>
+
+// Test based on: https://bugs.chromium.org/p/chromium/issues/detail?id=1419798#c5
+// Some allocators during deallocation may not call destructors and just reuse memory.
```

Comment:
Allocators don't destroy things internally, but the standard library may use an allocator to call a destructor. 

Does this mean to say "Some allocators don't call free from deallocate?" which would leave the memory in the  state string left it in?

If we really do mean "There's code that reuses the memory of live objects without destroying them", then that's exactly the kind of bug ASAN is meant to catch. That is not UB we should accept.

Reply:
This test case check a solution to the only known, potential issue with annotating all allocators, situations when memory is repurposed and destructors are not called.
Related issue: https://github.com/llvm/llvm-project/issues/60384

That's the only issue from past, I'm aware of (except bugs). That was addressed here: https://reviews.llvm.org/D145628
And it successfully works in `std::vector` and `std::deque`.

But is it UB? I remember that it's not clear.
Related discussion: https://reviews.llvm.org/D136765#4155262

As pointed out by @AntonBikineev 
> A program may end the lifetime of an object of class type without invoking the destructor, by reusing or releasing the storage as described above[.](https://eel.is/c++draft/basic.life#5.sentence-1)
> [...]
> [Note [4](https://eel.is/c++draft/basic.life#note-4): The correct behavior of a program often depends on the destructor being invoked for each object of class type[.](https://eel.is/c++draft/basic.life#5.sentence-4)
— end note]

Source:  https://eel.is/c++draft/basic.life#6

Therefore, it's not clear if it's UB. But even if it is, we want to make sure that it works that way, because people rely on this behavior.

Reply:
Added comments.

Reply:
> If we really do mean "There's code that reuses the memory of live objects without destroying them", then that's exactly the kind of bug ASAN is meant to catch. That is not UB we should accept.

This might by considered as UB in the past, but is not UB now per [CWG2523](https://cplusplus.github.io/CWG/issues/2523.html).

Reply:
Thx for the link! We do take care of that case, so no changes required, but I will add that link to the comment in the test.